rust/connlib/snownet/src/stun_binding.rs

```diff
@@ -104,33 +119,38 @@ impl StunBinding {
     }
 
     pub fn handle_timeout(&mut self, now: Instant) {
-        self.last_now = Some(now);
-
-        match self.state {
-            State::Initial => {
-                tracing::debug!(server = %self.server, "Sending new STUN request");
-            }
-            State::SentRequest { id, at } if at + STUN_TIMEOUT <= now => {
-                tracing::debug!(?id, "STUN request timed out, sending new one");
+        self.last_now = now;
+        self.backoff.clock.now = now;
+
+        let backoff = match self.state {
+            State::SentRequest { id, at, backoff } if at + backoff <= now => {
+                match self.backoff.next_backoff() {
+                    Some(backoff) => {
+                        tracing::debug!(?id, "STUN request timed out, sending new one");
+
+                        backoff
+                    }
+                    None => {
+                        tracing::info!(server = %self.server, "Giving up on attempting STUN binding");
+                        self.state = State::Failed;
+
+                        return;
+                    }
+                }
             }
             State::ReceivedResponse { at } if at + STUN_REFRESH <= now => {
                 tracing::debug!("Refreshing STUN binding");
+
+                self.backoff
+                    .next_backoff()
+                    .expect("to have initial backoff when we have received at least one response")
```

Comment:
What about resetting the backoff here instead of when we recieve the response? That way we make it more obvious why this should be `Some` and the behavior should be the same.

Reply:
I guess we could because they would just be "unecessary" resets and at least it is obvious then that this will always be `Some` here.

Reply:
Great idea!